### ISSUE TITLE
Error: MDB_CORRUPTED: Located page was wrong type

### ISSUE THREAD

**Issue by skyzyx** [parcel-bundler/parcel#8288]
## 🐛 bug report

> **IMPORTANT:** Both machines are running the same commit, and `npm install` was run separately on each machine. The `node_modules/` directory is `.gitignore`’d.

1. ✅ Running `parcel build` on an Intel-based MacBook Pro (16-inch, 2019) builds successfully.

2. ❌ Running `parcel build` on an Apple Silicon-based Mac Studio (2022) results in the error message:

    ```plain
    Error: MDB_CORRUPTED: Located page was wrong type
    ````

### package.json

```json
{
  "browserslist": "> 0.5%, last 2 versions, not dead",
  "source": [
    "layouts/index.html",
    "assets/css/index.css",
    "assets/js/_core.js"
  ],
  "targets": {
    "default": {
      "distDir": "../../dist",
      "publicUrl": "/static"
    }
  },
  "scripts": {
    "start": "make clean -C ../../ && TAILWIND_MODE=watch NODE_ENV=development npm-run-all pre:twcss --parallel dev:*",
    "lint": "npm-run-all eslint markdownlint",
    "eslint": "eslint --fix assets/js/_*.js",
    "markdownlint": "markdownlint --fix '**/*.md' --ignore 'node_modules'",
    "build": "make clean -C ../../ && NODE_ENV=production npm-run-all pre:twcss --parallel prod:*",
    "pre:twcss": "tailwindcss -i ./assets/css/_tailwind.css -o ./assets/css/index.css --jit",
    "dev:twcssw": "tailwindcss -i ./assets/css/_tailwind.css -o ./assets/css/index.css --jit -w",
    "dev:parcel": "parcel watch --detailed-report=10",
    "prod:twcss": "tailwindcss -i ./assets/css/_tailwind.css -o ./assets/css/index.css --jit --minify",
    "prod:parcel": "rm -f assets/js/index.js && parcel build --detailed-report=10 --no-cache --log-level=verbose && cp -vf ../../dist/assets/js/_core.js assets/js/index.js"
  },
  "dependencies": {
    "firecracker.js": "^1.0.5"
  },
  "devDependencies": {
    "@hyqo/parcel-reporter-asset-map": "^1.0.3",
    "@parcel/bundler-default": "^2.6.2",
    "@parcel/compressor-brotli": "^2.6.2",
    "@parcel/compressor-gzip": "^2.6.2",
    "@parcel/config-default": "^2.6.2",
    "@parcel/namer-default": "^2.6.2",
    "@parcel/optimizer-esbuild": "^2.6.2",
    "@parcel/packager-css": "^2.6.2",
    "@parcel/packager-html": "^2.6.2",
    "@parcel/packager-raw": "^2.6.2",
    "@parcel/reporter-cli": "^2.6.2",
    "@parcel/resolver-default": "^2.6.2",
    "@parcel/resolver-glob": "^2.6.2",
    "@parcel/transformer-html": "^2.6.2",
    "@parcel/transformer-js": "^2.6.2",
    "@parcel/transformer-postcss": "^2.6.2",
    "@parcel/transformer-raw": "^2.6.2",
    "@parcel/transformer-vue": "^2.6.2",
    "@tailwindcss/forms": "^0.5.2",
    "@tailwindcss/typography": "^0.5.2",
    "autoprefixer": "^10.4.7",
    "cssnano": "^5.1.12",
    "eslint-config-prettier": "^8.5.0",
    "npm-run-all": "^4.1.5",
    "parcel": "^2.6.2",
    "parcel-reporter-clean-dist": "^1.0.4",
    "parcel-resolver-ignore": "^2.1.3",
    "postcss": "^8.4.14",
    "postcss-cli": "^10.0.0",
    "postcss-custom-properties": "^12.1.8",
    "postcss-import": "^14.1.0",
    "postcss-url": "^10.1.3",
    "posthtml-doctype": "^1.1.1",
    "prettier": "^2.7.1",
    "prettier-plugin-go-template": "^0.0.13",
    "prettier-plugin-tailwindcss": "^0.1.11",
    "snyk": "^1.963.0",
    "tailwindcss": "^3.1.4"
  },
  "parcelIgnore": [
    "{{*.*}}",
    "/images/*.*"
  ]
}
```

### .parcelrc

```json
{
  "extends": "@parcel/config-default",
  "bundler": "@parcel/bundler-default",
  "transformers": {
    "*.html": [
      "@parcel/transformer-html"
    ],
    "*.js": [
      "@parcel/transformer-js"
    ],
    "*.css": [
      "@parcel/transformer-postcss",
      "@parcel/transformer-css"
    ],
    "*": [
      "@parcel/transformer-raw"
    ]
  },
  "namers": [
    "…",
    "@parcel/namer-default"
  ],
  "optimizers": {
    "*.html": [
      "@parcel/optimizer-htmlnano"
    ],
    "*.js": [
      "@parcel/optimizer-esbuild"
    ],
    "*.css": [
      "@parcel/optimizer-css"
    ]
  },
  "packagers": {
    "*.html": "@parcel/packager-html",
    "*.css": "@parcel/packager-css",
    "*.js": "@parcel/packager-js",
    "*": "@parcel/packager-raw"
  },
  "resolvers": [
    "…",
    "parcel-resolver-ignore",
    "@parcel/resolver-glob"
  ],
  "compressors": {
    "*.{js,html,css}": [
      "…",
      "@parcel/compressor-gzip",
      "@parcel/compressor-brotli"
    ]
  },
  "reporters": [
    "…",
    "@parcel/reporter-cli",
    "parcel-reporter-clean-dist",
    "@hyqo/parcel-reporter-asset-map"
  ]
}
```

## 🤔 Expected Behavior

`parcel build` would complete successfully on both CPU architectures.

### Intel-based MacBook Pro (16-inch, 2019)

```plain
$ npm run build

> build
> make clean -C ../../ && NODE_ENV=production npm-run-all pre:twcss --parallel prod:*

make: Entering directory '/Library/WebServer/Documents/fpb.com'

=====> Cleaning Hugo build artifacts…
rm -Rf ./public

=====> Cleaning npm…
rm -Rf ./node_modules
make: Leaving directory '/Library/WebServer/Documents/fpb.com'

> pre:twcss
> tailwindcss -i ./assets/css/_tailwind.css -o ./assets/css/index.css --jit


Done in 313ms.

> prod:parcel
> rm -f assets/js/index.js && parcel build --detailed-report=10 --no-cache --log-level=verbose && cp -vf ../../dist/assets/js/_core.js assets/js/index.js


> prod:twcss
> tailwindcss -i ./assets/css/_tailwind.css -o ./assets/css/index.css --jit --minify

⠋ Building index.js…

✨ Built in 2.22s

../../dist/assets/css/index.css      14.87 KB     27ms
├── assets/css/index.css             14.83 KB    1.28s
└── Code from unknown sourcefiles        38 B      0ms

../../dist/assets/js/_core.js          5.5 KB    544ms
├── ../../dist/assets/js/_core.js     5.47 KB      0ms
└── Code from unknown sourcefiles        35 B      0ms

../../dist/layouts/index.html           409 B    290ms
└── layouts/index.html                  427 B    343ms
✨ Built in 2.22s

../../dist/assets/css/index.css      14.87 KB     27ms
├── assets/css/index.css             14.83 KB    1.28s
└── Code from unknown sourcefiles        38 B      0ms

../../dist/assets/js/_core.js          5.5 KB    544ms
├── ../../dist/assets/js/_core.js     5.47 KB      0ms
└── Code from unknown sourcefiles        35 B      0ms

../../dist/layouts/index.html           409 B    290ms
└── layouts/index.html                  427 B    343ms
console: Asset map 🗺: ../../dist/asset-map.json
console: Asset map 🗺: ../../dist/asset-map.json
'../../dist/assets/js/_core.js' -> 'assets/js/index.js'
```

### Apple Silicon-based Mac Studio (2022)

```plain
npm run build

> build
> make clean -C ../../ && NODE_ENV=production npm-run-all pre:twcss --parallel prod:*

make: Entering directory '/Library/WebServer/Documents/fpb.com'

=====> Cleaning Hugo build artifacts…
rm -Rf ./public

=====> Cleaning npm…
rm -Rf ./node_modules
make: Leaving directory '/Library/WebServer/Documents/fpb.com'

> pre:twcss
> tailwindcss -i ./assets/css/_tailwind.css -o ./assets/css/index.css --jit


Done in 167ms.

> prod:parcel
> rm -f assets/js/index.js && parcel build --detailed-report=10 --no-cache --log-level=verbose && cp -vf ../../dist/assets/js/_core.js assets/js/index.js


> prod:twcss
> tailwindcss -i ./assets/css/_tailwind.css -o ./assets/css/index.css --jit --minify

Error: MDB_CORRUPTED: Located page was wrong type
    at /Library/WebServer/Documents/fpb.com/themes/fpb/node_modules/lmdb/dist/index.cjs:2149:16
    at LMDBStore.transactionSync (/Library/WebServer/Documents/fpb.com/themes/fpb/node_modules/lmdb/dist/index.cjs:910:17)
    at new LMDBStore (/Library/WebServer/Documents/fpb.com/themes/fpb/node_modules/lmdb/dist/index.cjs:2148:10)
    at Object.open (/Library/WebServer/Documents/fpb.com/themes/fpb/node_modules/lmdb/dist/index.cjs:2327:35)
    at new LMDBCache (/Library/WebServer/Documents/fpb.com/themes/fpb/node_modules/@parcel/cache/lib/LMDBCache.js:81:34)
    at resolveOptions (/Library/WebServer/Documents/fpb.com/themes/fpb/node_modules/@parcel/core/lib/resolveOptions.js:125:168)
    at async Parcel._init (/Library/WebServer/Documents/fpb.com/themes/fpb/node_modules/@parcel/core/lib/Parcel.js:218:27)
    at async Parcel.run (/Library/WebServer/Documents/fpb.com/themes/fpb/node_modules/@parcel/core/lib/Parcel.js:273:7)
    at async run (/Library/WebServer/Documents/fpb.com/themes/fpb/node_modules/parcel/lib/cli.js:376:7) {
  code: -30796
}
ERROR: "prod:parcel" exited with 1.
```

## 💁 Possible Solution

Parcel appears to rely on <https://github.com/kriszyp/lmdb-js>. Looking at the latest release (as of this moment, <https://github.com/kriszyp/lmdb-js/releases/tag/v2.5.3>), it appears to be building pre-bundled packages for

* prebuild-alpine.tar (9.24 MB)
* prebuild-darwin.tar (4.09 MB)
* prebuild-linux-arm.tar (4.37 MB)
* prebuild-linux.tar (3.33 MB)

Looking at these releases, I notice two things:

1. `darwin` is listed, but not `darwin-arm` — like the Linux releases.

2. The `darwin` version is roughly the same size as each linux version, instead of being double-sized, which suggests that the `darwin` version is not a _Universal Binary_ containing both CPU architectures inside a single binary.

## 🔦 Context

1. Trying to compile my CSS and JS for a relatively simple, small website I'm working on.

2. Since replacing my personal Intel-based Mac with a personal Apple Silicon-based Mac, now my only Intel-based Mac is my work laptop, and I try to avoid doing personal work on my corporate machine.

## 💻 Code Sample

I believe I've provided everything necessary, above.

## 🌍 Your Environment

Both machines have the same software specs.

| Software         | Version(s) |
| ---------------- | ---------- |
| Parcel           | v2.6.2 |
| Node             | v18.4.0 |
| npm/Yarn         | v8.12.1 |
| Operating System | macOS 12.5 Beta (21G5056b) |

**Comment by skyzyx** [comment]
Cross-posting for visibility: https://github.com/kriszyp/lmdb-js/issues/185

**Comment by kriszyp** [comment]
Just FYI on the builds, MacOS supports cross-compilation, so there is just one tar needed to archive both the arm64 and x64 builds, but they are ultimately distributed separately (and you can open the tar archive and verify they are different sizes, and no I am not using universal binaries, the binaries for x64 and arm64 are distinct). You can verify the binary's existence at @lmdb/lmdb-darwin-arm64. And I would guess that there are probably [hundreds or thousands of devs](https://www.npmjs.com/package/@lmdb/lmdb-darwin-arm64) using parcel/lmdb-js on M1's, so it is tested.

But all that being said, I certainly do not doubt you have a unique combination of packages/config that is legitimately triggering this error. Unfortunately, I do not have access to an M1 myself (I have had friends test/verify lmdb-js for me in the past), and I don't think github actions has any M1 runners yet. I do have a CI test for emulating Arm7 with QEMU/Docker, and I could see if I could possibly reproduce your issue there (in case it is ARM-specific). It seems you have left pretty clear steps, but I think I would also need files from the repo you are using? Also, based on your report it looks like this is crashing pretty quickly? And even from a fresh directory?

I also wonder if this is possibly related to using a beta version of MacOS? If you have time, does cloning https://github.com/kriszyp/lmdb-js, `git checkout v2.5`, `npm install`, and `npm run test` succeed?

**Comment by skyzyx** [comment]
[@kriszyp](https://github.com/kriszyp) wrote:

> But all that being said, I certainly do not doubt you have a unique combination of packages/config that is legitimately triggering this error. […] It seems you have left pretty clear steps, but I think I would also need files from the repo you are using?

Here is my JavaScript — which I didn't include before — that I'm trying to compile with Parcel. I posted the rest of my `package.json` previously. It's just some close-to-the-metal DOM manipulation and old-school event delegation. It's also not very much JavaScript. I count 26 lines.

```javascript
import Firecracker from 'firecracker.js/dist/index';

const $ = Firecracker.DQuery,
  listen = Firecracker.Delegate,
  svgs = $('#mobile-menu-button svg');

$(document.body).on('click', (e) => {
  // Mobile menu
  listen('#mobile-menu-button', () => {
    // Change icon
    svgs.forEach((el) => {
      el.toggle('hidden');
      el.toggle('block');
    });

    // Show menu
    $('#mobile-menu')[0].toggle('hidden');
  })(e);

  // Chapter toggle
  listen('#chapter-toggle', () => {
    document.querySelectorAll('details').forEach((a) => {
      a.open = !a.open;
    });
  })(e);
});
```

----

> Also, based on your report it looks like this is crashing pretty quickly? And even from a fresh directory?

By "fresh directory", I'm assuming you mean that it's essentially just `package.json`, `package-lock.json`, the Parcel definitions, and `node_modules/` — in which case, yes.

If I re-run `npm start` with `time`, I get:

```plain
$ time npm start

> start
> make clean -C ../../ && TAILWIND_MODE=watch NODE_ENV=development npm-run-all pre:twcss --parallel dev:*

make: Entering directory '/Library/WebServer/Documents/fpb.com'

=====> Cleaning Hugo build artifacts...
rm -Rf ./public

=====> Cleaning npm...
rm -Rf ./node_modules
make: Leaving directory '/Library/WebServer/Documents/fpb.com'

> pre:twcss
> tailwindcss -i ./assets/css/_tailwind.css -o ./assets/css/index.css --jit


Done in 156ms.

> dev:parcel
> parcel watch --detailed-report=10


> dev:twcssw
> tailwindcss -i ./assets/css/_tailwind.css -o ./assets/css/index.css --jit -w

Error: MDB_CORRUPTED: Located page was wrong type
    at /Library/WebServer/Documents/fpb.com/themes/fpb/node_modules/lmdb/dist/index.cjs:2149:16
    at LMDBStore.transactionSync (/Library/WebServer/Documents/fpb.com/themes/fpb/node_modules/lmdb/dist/index.cjs:910:17)
    at new LMDBStore (/Library/WebServer/Documents/fpb.com/themes/fpb/node_modules/lmdb/dist/index.cjs:2148:10)
    at Object.open (/Library/WebServer/Documents/fpb.com/themes/fpb/node_modules/lmdb/dist/index.cjs:2327:35)
    at new LMDBCache (/Library/WebServer/Documents/fpb.com/themes/fpb/node_modules/@parcel/cache/lib/LMDBCache.js:81:34)
    at resolveOptions (/Library/WebServer/Documents/fpb.com/themes/fpb/node_modules/@parcel/core/lib/resolveOptions.js:125:168)
    at async Parcel._init (/Library/WebServer/Documents/fpb.com/themes/fpb/node_modules/@parcel/core/lib/Parcel.js:218:27)
    at async Parcel.watch (/Library/WebServer/Documents/fpb.com/themes/fpb/node_modules/@parcel/core/lib/Parcel.js:315:7)
    at async run (/Library/WebServer/Documents/fpb.com/themes/fpb/node_modules/parcel/lib/cli.js:351:9) {
  code: -30796
}

ERROR: "dev:parcel" exited with 1.

real	0m1.612s
user	0m1.396s
sys	0m0.431s
```

_1.612 seconds_.

----

> I also wonder if this is possibly related to using a beta version of macOS?

I mean, the same beta build (`21G5056b`) on the Intel Mac works without issue. `¯\_(ツ)_/¯`

----

> If you have time, does cloning https://github.com/kriszyp/lmdb-js, `git checkout v2.5`, `npm install`, and `npm run test` succeed?

Each command is its own code block for readability. In the end, I get a segfault when running the tests.

```plain
$ git clone git@github.com:kriszyp/lmdb-js.git
Cloning into 'lmdb-js'...
remote: Enumerating objects: 250798, done.
remote: Counting objects: 100% (909/909), done.
remote: Compressing objects: 100% (370/370), done.
remote: Total 250798 (delta 562), reused 783 (delta 464), pack-reused 249889
Receiving objects: 100% (250798/250798), 46.71 MiB | 20.35 MiB/s, done.
Resolving deltas: 100% (189018/189018), done.
```

```plain
$ cd lmdb-js/
/Library/WebServer/Documents/lmdb-js
```

```plain
$ git checkout v2.5
branch 'v2.5' set up to track 'origin/v2.5' by rebasing.
Switched to a new branch 'v2.5'
```

```plain
$ npm install
npm WARN deprecated har-validator@5.1.5: this library is no longer supported
npm WARN tarball tarball data for escape-goat@https://registry.npmjs.org/escape-goat/-/escape-goat-2.1.1.tgz (sha512-8/uIhbG12Csjy2JEW7D9pHbreaVaS/OpN3ycnyvElTdwM5n6GY6W6e2IPemfvGZeUMqZ9A/3GqIZMgKnBhAw/Q==) seems to be corrupted. Trying again.
npm WARN deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142

> lmdb@2.5.3 install
> node-gyp-build-optional-packages


> lmdb@2.5.3 prepare
> rollup -c


node-index.js → dist/index.cjs...
(!) Unresolved dependencies
https://rollupjs.org/guide/en/#warning-treating-module-as-external-dependency
module (imported by node-index.js)
v8 (imported by node-index.js)
events (imported by index.js)
os (imported by index.js)
fs (imported by index.js)
msgpackr (imported by index.js)
weak-lru-cache (imported by index.js)
ordered-binary (imported by index.js)
path (imported by native.js)
url (imported by native.js)
node-gyp-build-optional-packages (imported by native.js)
(!) Unused external imports
join imported from external module "path" but never used in "native.js"
(!) Mixing named and default exports
https://rollupjs.org/guide/en/#outputexports
The following entry modules are using named and default exports together:
node-index.js

Consumers of your bundle will have to use chunk['default'] to access their default export, which may not be what you want. Use `output.exports: 'named'` to disable this warning
created dist/index.cjs in 120ms

added 411 packages, and audited 412 packages in 24s

51 packages are looking for funding
  run `npm fund` for details

5 moderate severity vulnerabilities

To address all issues (including breaking changes), run:
  npm audit fix --force

Run `npm audit` for details.
```

```plain
$ npm run test

> lmdb@2.5.3 test
> mocha test/**.test.js --recursive && npm run test:types



  lmdb-js
    Basic use
      ✔ will not open non-existent db with create disabled
      -
      ✔ zero length values
      ✔ query of keys
      ✔ reverse query range
      ✔ more reverse query range
      ✔ clear between puts
      ✔ string
      ✔ string with version
      ✔ string with version branching
      ✔ repeated ifNoExists (3373ms)
      ✔ string with compression and versions
      ✔ repeated compressions
      ✔ forced compression due to starting with 255
      ✔ store objects
      ✔ store binary
      ✔ writes batch with callback
      ✔ conditional put
      - trigger sync commit
      ✔ should iterate over query
      ✔ should iterate over query, across transactions
      ✔ should break out of query
      ✔ getRange with arrays
      ✔ should iterate over query with offset/limit
      ✔ should handle open iterators and cursor renewal
      ✔ should iterate over dupsort query, with removal
      ✔ should iterate over ordered-binary dupsort query with start/end
      ✔ should count ordered-binary dupsort query with start/end
      ✔ should reverse iterate ordered-binary dupsort query with start/end
      ✔ doesExist
      ✔ should iterate over keys without duplicates
      ✔ big keys
      ✔ prefetch
      ✔ invalid key
      ✔ put options (sync)
      ✔ async transactions
      ✔ child transaction in sync transaction
      ✔ async transaction with interrupting sync transaction default order
      ✔ multiple async mixed
      ✔ big child transactions
      ✔ handle write transaction with hanging cursors
      ✔ mixed batches (48ms)
      ✔ levelup style callback
      ✔ batch operations
      ✔ batch array
      ✔ read and write with binary encoding (157ms)
      ✔ read and write with binary encoding of key and value
      - large txn
/var/folders/1j/l_wsg4bs739f60tkjvvkz0800000gn/T/test658297338.sh: line 2: 37600 Segmentation fault: 11  mocha test/**.test.js --recursive
```

**Comment by skyzyx** [comment]
In the test run above, I see a few weird things:

* "escape-goat […] seems to be corrupted. Trying again."

* "(!) Unresolved dependencies"

* And then, of course, the segmentation fault.

```bash
$ ./node_modules/.bin/parcel --version
2.6.2

$ node --version
v18.6.0

$ npm --version
8.13.2

$ osqueryi --json "SELECT name, version, platform, arch, build FROM os_version;" | jq -Mr '.[] | "\(.name) \(.version) (\(.build)) (\(.platform)/\(.arch))"'
macOS 12.5 (21G69) (darwin/arm64)
```

**Comment by kriszyp** [comment]
@skyzyx Thank you for the detailed report, this is certainly very helpful, and it is good to know a segfault is reproducible in the unit tests. I will move this conversation to the lmdb-js issue.

**Comment by skyzyx** [comment]
While trying to narrow down a smaller repro case using a small Alpine Linux arm64 container (to see if it's more related to macOS or arm64), I was able to trigger the issue, then was able to _un-trigger_ the issue last night (i.e., got a successful Parcel build).

Something I'm doing in my full project is triggering this error, that I wasn't replicating in the repro test case. Since I was able to reproduce the error in the `linux/arm64` container, I'm led to believe that the issue is more related to arm64 than macOS (including the beta builds).

I'm going to continue poking at it this evening to see if I can track down which part of my project is the trigger — see if I can narrow the root cause down even further.

Keeping the ball…

**Comment by skyzyx** [comment]
Once I deleted the `.parcel-cache` directory, the problem went away. I think that when I switched machines (from Intel to Apple Silicon), and I restored my project from backup, I brought over the `.parcel-cache` directory which had been built on the wrong CPU architecture.

A massive **thank you** to @kriszyp for helping me get to the bottom of this issue. It's obvious that you _give a shit_, which is a breath of fresh air on the internet in 2022. Thank you kind stranger.
